### DISCUSSION TITLE
Preview Release with XRI3 Support: Core v4.0.0-pre.1, Input v4.0.0-pre.1, Spatialmanipulation v4.0.0-pre.1, Uxcomponents v4.0.0-pre.1, Uxcomponents.noncanvas v4.0.0-pre.1, Uxcore v4.0.0-pre.1

### DISCUSSION THREAD

**Discussion by shaynie** [MixedRealityToolkit/MixedRealityToolkit-Unity#894] [Releases]
![ReleaseImage](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/9615036/bea24c69-25e7-4f69-953d-91dca9ddc233)

This preview release updates MRTK3 to support Unity's latest [XR Interaction Toolkit](https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@3.0/manual/index.html) (XRI) version 3.0.  Please try this preview release and open any new issues at https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues.

There are set of known issues with these preview packages. Fixes for these issues will be made available in future releases.  The known issues are:

  - https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/878
  - https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/877
  - https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/876 
  - https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/880
  - https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/885
 

Progress of officially supporting XRI3 can be found at https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/778.

# Changelog for MRTK
The format is based on [Keep a Changelog] - (https://keepachangelog.com/en/1.1.0/).

### Fixed  
* Reduced package description to support for UPM package publishing in the Unity Asset Store.  
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519)
 
### Fixed  
* Reduced package description to support for UPM package publishing in the Unity Asset Store. 
* Fixing .asmdef for Editor Unit Tests, so it now contains proper `defineConstraints`.  
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519)
 
### Added  
* Added ITrackedInteractor to represent interactor with parent pose backed by a tracked input device.  
### Changed  
* Updated package com.unity.xr.interaction.toolkit to 3.0.3 
* Updated InteractorHandednessExtensions.  
### Removed  
* Removed obsolete HandednessExtensions::IsRight method. 
* Removed obsolete HandednessExtensions::IsLeft method. 
* Removed obsolete HandsUtils::GetSubsystem method. 
* Removed obsolete PlayspaceUtilities.ReferenceTransform field. 
* Removed obsolete XRSubsystemHelpers::GetAllRunningSubsystemsNonAlloc method.  
### Deprecated  
* ControllerLookup marked as Obsolete.  
### Fixed  
* Fixed UPM package validation so that it ignores errors caused when the test runner is not part of the MRTK publisher account. [PR 
#775](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/775/)  
### Fixed  
* Fixed missing [CanEditMultipleObject] attributes as per Bug 573 [PR 
#698](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/698)  
### Added  
* StabilizedRay constructor with explicit position and direction half life values. [PR 
#625](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/625) 
* Added IsProximityHovered property of type TimedFlag to detect when a button starts being hovered or on interactor proximity and when it stops being hovered or on proximity of any interactor. [PR 
#611](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/611) 
* Adding ProximityHover events (Entered & Exited) to PressableButton class. [PR 
#611](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/611)   
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519) 
* Fix warning and event triggered on disabled StatefulInteractable after changing speech settings [PR 
#591](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/591) [PR 
#608](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/608)
 
### Fixed  
* Fixed missing [CanEditMultipleObject] attributes as per Bug 573 [PR 
#698](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/698) 
* Reduced package description to support for UPM package publishing in the Unity Asset Store.  
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519)
 
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519)
 
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519)
 
### Added  
* SpatialMouseInputTestsForControllerlessRig Unity-tests. 
* BasicInputTestsForControllerlessRig Unity-tests. 
* Ported BasicInputTests so that they also test the new controllerless prefabs. 
* Ported FuzzyGazeInteractorTests so that they also test the new controllerless prefabs. 
* MRTKRayInteractorVisualsTestsForControllerlessRig Unity-tests. 
* InteractionModeManagerTestsForControllerlessRig Unity-tests. 
* FuzzyGazeInteractorTestsForControllerlessRig Unity-tests. 
* TrackedPoseDriverLookup as the XRI3+ equivalent of ControllerLookup. 
* TrackedPoseDriverWithFallback as the XRI3+ equivalent of ActionBasedControllerWithFallbacks. 
* Controllerless version of MRTK XR Rig prefab. 
* Controllerless version of MRTK LeftHand Controller prefab. 
* Controllerless version of MRTK RightHand Controller prefab. 
* Controllerless version of MRTK Gaze Controller prefab. 
* Controllerless version of MRTK Interaction Manager prefab. 
* Added ModeManagerdRoot field to interactors to hold a reference to parent GameObject.  
### Changed  
* Updated package com.unity.xr.interaction.toolkit to 3.0.3 
* Updated BaseRuntimeInputTests logic to handle both deprecated XRController and new controllerless actions. 
* Updated GazePinchInteractor logic to handle both deprecated XRController and new controllerless actions. 
* Updated PokeInteractor logic to handle both deprecated XRController and new controllerless actions. 
* Updated MRTKRayInteractor logic to handle both deprecated XRController and new controllerless actions. 
* Updated FlatScreenModeDetector logic to handle both deprecated XRController and new controllerless actions. 
* Updated ObjectManipulator so to not rely on obsolete XRI controllers. 
* Moved the Gaze Interactor TrackedPoseDriver to parent GameObject so that all controller prefabs have the same structure. 
* Moved HandModel script from Experimental\XRI3 to Controllers\ 
* Renamed MRTK XR Rig prefab as Obsolete MRTK XR Rig. 
* Renamed MRTK LeftHand Controller prefab as Obsolete MRTK LeftHand Controller. 
* Renamed MRTK RightHand Controller prefab as Obsolete MRTK RightHand Controller. 
* Renamed MRTK Gaze Controller prefab as Obsolete MRTK Gaze Controller. 
* Renamed MRTK Interaction Manager prefab as Obsolete MRTK Interaction Manager. 
* Added ITrackedInteractor interface to GazePinchInteractor class. 
* Added ITrackedInteractor interface to HandJointInteractor class. 
* Added ITrackedInteractor interface to PokeInteractor class. 
* Added ITrackedInteractor interface to MRTKRayInteractor class. 
* Updated new controllerless rig to use HandPoseDrive and PinchInputReader to support devices without a Hand Interaction profile.  
### Deprecated  
* ActionBasedControllerWithFallbacks marked as Obsolete. 
* ArticulatedHandController marked as Obsolete.  
### Removed  
* Removed obsolete ArticulatedHandController.HandsAggregatorSubsystem field. 
* Removed obsolete MRTKRayInteractor.HandsAggregatorSubsystem field. 
* Removed obsolete ControllerSimulationSettings.InputActionReference field. 
* Removed obsolete SyntheticsHandsSubsystem::GetNeutralPose method. 
* Removed obsolete SyntheticsHandsSubsystem::SetNeutralPose method. 
* Removed obsolete SyntheticsHandsSubsystem::GetSelectionPose method. 
* Removed obsolete SyntheticsHandsSubsystem::SetSelectionPose method. 
* Removed obsolete SyntheticsHandsSubsystem::GetNeutralPose method. 
* Removed obsolete SyntheticsHandsSubsystem::SetNeutralPose method. 
* Removed obsolete FollowJoint.migratedSuccessfully field. 
* Removed obsolete FollowJoint.hand field. 
* Removed obsolete FollowJoint.Joint field. 
* Removed obsolete FollowJoint::OnAfterDeserialize method. 
* Removed obsolete HandBasedPoseSource.HandsAggregator field. 
* Removed obsolete ControllerVisualizer.HandsAggregator field. 
* Removed no longer needed Experimental\XRI3 folder  
### Fixed  
* InputSimulator execution order so that it executes before InteractionManager.  
### Fixed  
* Fixed missing [CanEditMultipleObject] attributes as per Bug 573 [PR 
#698](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/698) 
* Add logic to account for a bound but untracked interaction profile. [PR 
#704](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/704) 
* Reduced package description to support for UPM package publishing in the Unity Asset Store. 
* Ensures the simulated input sources hold their state (including gestures) when their toggle state is locked on. [PR 
#705](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/705)  
### Added  
* Added an alternative Line of Sight (LOS), with angular offset, hand ray pose source. [PR 
#625](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/625) 
* Added IsProximityHovered property of type TimedFlag to detect when a button starts being hovered or on interactor proximity and when it stops being hovered or on proximity of any interactor. [PR 
#611](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/611)  
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519) 
* Fix the fallback controllers being backwards [PR 
#636](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/636) 
* Fix empty SpeechRecognitionKeyword breaking all speech keyword system [PR 
#612](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/612) [PR 
#614](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/614)
 
### Added  
* SolverTapToPlaceTestsForControllerlessRig Unity-tests. 
* Ported SolverTapToPlaceTests so that they also test the new controllerless prefabs. 
* Updated TapToPlace logic to handle both deprecated XRController and new controllerless actions. 
* Updated HandConstraintPalmUp logic to handle both deprecated XRController and new controllerless actions. 
* Updated Solver logic to handle both deprecated XRController and new controllerless actions.  
### Changed  
* Updated package com.unity.xr.interaction.toolkit to 3.0.3  
### Fixed  
* Fixed tap to place `StartPlacement()` when called just after instantiation of the component. [PR 
#785](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/785)  
### Added  
* Made bounds control overridable for custom translation, scaling and rotation logic using manipulation logic classes. [PR 
#722](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/722)  
### Fixed  
* Added null check and index check when hiding colliders on BoundsHandleInteractable. [PR 
#730](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/730)  
### Added  
* ObjectManipulator's ManipulationLogic observes XRSocketInteractor, XRI v2.3.0. [PR 
#567](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/567)  
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519)
 
### Added  
* Added unified font atlas and updated corresponding fonts and their materials. [PR 
#700](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/700)  
### Added  
* Added SimpleEmptyButton (Experimental) and SimpleActionButton (Experimental) prefabs plus corresponding (Experimental) assets and demoscene. [PR 
#635](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/635)  
### Fixed        
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519) 
* Fixing shaders of a couple of TMPro fonts [PR 
#696](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/696) 
* Set the Data packages's CooperHewitt-BoldItalic SDK shader to the Graphic Tools TMPro shader [PR 
#696](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/696)
 
### Fixed  
* Fixing .asmdef for Editor Unit Tests, so it now contains proper `defineConstraints`.  
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519)
 
### Changed  
* Updated package com.unity.xr.interaction.toolkit to 3.0.3  
### Added  
* Added proximity-hover dynamic Frontplates functionality (Moved from Experimental to Release). [PR 
#712](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/712) 
* Added unified font atlas and updated corresponding fonts and their materials. [PR 
#700](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/700)  
### Fixed  
* Reduced package description to support for UPM package publishing in the Unity Asset Store. 
* Changed PressableButton_Custom_Cylinder using Default-Material to MRTK_Standard_White instead. [PR 
#740](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/740)  
### Added  
* Added Empty, Action, Action with Checkbox, and ToggleSwitch Experimental buttons prefabs [PR 
#649](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/649) 
* Added corresponding MenuItems for use in Editor Tools [PR 
#649](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/649) 
* Added CanvasUITearsheetDynamicFrontplate scene [PR 
#649](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/649) 
* Added SimpleButton + demo scene [PR 
#635](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/635)  
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519) 
* Fixed Bug 
#643 - Experimental SimpleEmptyButton and SimpleActionButton prefabs have missing "See It Say It Label" GameObject reference [PR 
#646](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/646) 
* Auto disable button's colliders when `StatefulInteractable` is disabled [PR 
#626](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/626)
 
### Changed  
* Updated package com.unity.xr.interaction.toolkit to 3.0.3  
### Fixed  
* Reduced package description to support for UPM package publishing in the Unity Asset Store.  
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519)
 
### Changed  
* Updated package com.unity.xr.interaction.toolkit to 3.0.3  
### Removed  
* Removed LegacyDialog/Dialog files. 
* Removed LegacyDialog/DialogButton files. 
* Removed LegacyDialog/DialogButtonContext files. 
* Removed LegacyDialog/DialogButtonHelpers files. 
* Removed LegacyDialog/DialogButtonTypes files. 
* Removed LegacyDialog/DialogProperty files. 
* Removed LegacyDialog/DialogShell files. 
* Removed LegacyDialog/DialogState files. 
* Removed LegacyDialog/README files. 
* Removed obsolete Slider.SliderValue fields.  
### Fixed  
* Fixed missing [CanEditMultipleObject] attributes as per Bug 573 [PR 
#698](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/698) 
* Reduced package description to support for UPM package publishing in the Unity Asset Store.  
### Added  
* Added Experimental Buttons with dynamic Frontplate and demo CanvasUITearsheetDynamicFrontplate scene. [PR 
#649](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/649)  
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519) 
* Fixed some buttons in nonnative keyboard not having sound [PR 
#648](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/648)
 
### Fixed  
* Reduced package description to support for UPM package publishing in the Unity Asset Store.  
### Fixed  
* Fixed support for UPM package publishing in the Unity Asset Store. [PR 
#519](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/519)

<hr /><em>This discussion was created from the release <a href='https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/releases/tag/core-v4.0.0-pre.1'>Core v4.0.0-pre.1, Input v4.0.0-pre.1, Spatialmanipulation v4.0.0-pre.1, Uxcomponents v4.0.0-pre.1, Uxcomponents.noncanvas v4.0.0-pre.1, Uxcore v4.0.0-pre.1</a>.</em>

**Comment by XRWebland** [comment]
is there any updates for when 4.0 with XRI3 will be released officially instead of a pre release?

**Reply by keveleigh** [reply]
I can't commit to a date, but I'm actively working on this!
